AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStore/src/DefaultEncryptedKeyStore.dfy

```diff
@@ -0,0 +1,720 @@
+          && Seq.Last(ddbClient.History.TransactWriteItems).input
+             //= aws-encryption-sdk-specification/framework/key-store/dynamodb-encrypted-key-store.md#writenewbranchkeyversiontokeystore
+             //= type=implication
+             //# The operation MUST call Amazon DynamoDB TransactWriteItems with a request constructed as follows:
+             == DDB.TransactWriteItemsInput(
+                  TransactItems := [
+                    //= aws-encryption-sdk-specification/framework/key-store/dynamodb-encrypted-key-store.md#writenewbranchkeyversiontokeystore
+                    //= type=implication
+                    //#- PUT:
+                    //#  - Item: A [record formatted item](#record-format) constructed from the version input
+                    //#  - ConditionExpression: `attribute_not_exists(branch-key-id)`
+                    //#  - TableName: the configured Table Name
+                    CreateTransactWritePutItem(
+                      input.Version,
+                      ddbTableName,
+                      BRANCH_KEY_NOT_EXIST
+                    ),
+                    //= aws-encryption-sdk-specification/framework/key-store/dynamodb-encrypted-key-store.md#writenewbranchkeyversiontokeystore
+                    //= type=implication
+                    //#- PUT:
+                    //#  - Item: A [record formatted item](#record-format) constructed from the active input
+                    //#  - ConditionExpression: `attribute_exists(branch-key-id)`
+                    //#  - TableName: the configured Table Name
+                    CreateTransactWritePutItem(
+                      input.Active,
+                      ddbTableName,
+                      BRANCH_KEY_EXISTS
+                    )
+                  ]
+                )
+    {
+
+      :- Need(
+        && (forall k <- input.Version.EncryptionContext.Keys :: DDB.IsValid_AttributeName(k))
+        && (forall k <- input.Active.EncryptionContext.Keys :: DDB.IsValid_AttributeName(k))
+      , Types.KeyStoreException( message := ErrorMessages.ENCRYPTION_CONTEXT_EXCEEDS_DDB_LIMIT)
+      );
+
+      var items: DDB.TransactWriteItemList := [
+        CreateTransactWritePutItem(
+          input.Version,
+          ddbTableName,
+          BRANCH_KEY_NOT_EXIST
+        ),
+        CreateTransactWritePutItem(
+          input.Active,
+          ddbTableName,
+          BRANCH_KEY_EXISTS
+        )
+      ];
+
+      var transactRequest := DDB.TransactWriteItemsInput(
+        TransactItems := items,
+        ReturnConsumedCapacity := None,
+        ReturnItemCollectionMetrics := None,
+        ClientRequestToken := None
+      );
+
+      var transactWriteItemsResponse? := ddbClient.TransactWriteItems(transactRequest);
+      var BRANCH_KEY_ACTIVE_VERSION_FIELD :- transactWriteItemsResponse?
+      .MapFailure(e => Types.ComAmazonawsDynamodb(ComAmazonawsDynamodb := e));
```

Comment:
Issue/Blocking: I hope that Dafny will not let a constant field by modified, 
and strictly speaking this `BRANCH_KEY_ACTIVE_VERSION_FIELD` appears to be scoped to
only this method,
as compared to `Structure.BRANCH_KEY_ACTIVE_VERSION_FIELD`, 
but I still strongly object to this statement as it is.

```suggestion
      var _ :- transactWriteItemsResponse?
      .MapFailure(e => Types.ComAmazonawsDynamodb(ComAmazonawsDynamodb := e));
```

Reply:
changed. There is no way to update a const filed, but also this is not the same field.